File: Assignment quick sort v2/newQuick.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <utility>
#include <algorithm>
#include <chrono>

using namespace std;

typedef pair<int, string> Pair;
vector<Pair> elements;

void quickSort(int low, int high);
int partition(int low, int high);
void medianOfThree(int low, int high);

int main() {
    string inputFile = "dataset_5000000.csv";
    ifstream inFile(inputFile);
    string line;

    if (!inFile.is_open()) {
        cerr << "Cannot open file: " << inputFile << endl;
        return 1;
    }

    elements.reserve(5'000'000); // Reserve space for performance

    while (getline(inFile, line)) {
        stringstream ss(line);
        string numStr, text;

        if (getline(ss, numStr, ',') && getline(ss, text)) {
            try {
                int num = stoi(numStr);
                elements.emplace_back(num, text);
            } catch (...) {
                cerr << "Invalid line skipped: " << line << endl;
            }
        }
    }
    inFile.close();

    auto start = chrono::high_resolution_clock::now();
    quickSort(0, elements.size() - 1);
    auto end = chrono::high_resolution_clock::now();

    /*string outputFile = "quick_sort_" + to_string(elements.size()) + ".csv";
    ofstream outFile(outputFile);

    for (const auto& p : elements) {
        outFile << p.first << "," << p.second << "\n";
    }
    outFile.close();*/

    chrono::duration<double> elapsed = end - start;
    cout << "Sorting time: " << elapsed.count() << " seconds" << endl;

    return 0;
}

// Median-of-three pivot strategy
void medianOfThree(int low, int high) {
    int mid = low + (high - low) / 2;

    if (elements[high].first < elements[low].first)
        swap(elements[low], elements[high]);
    if (elements[mid].first < elements[low].first)
        swap(elements[mid], elements[low]);
    if (elements[high].first < elements[mid].first)
        swap(elements[mid], elements[high]);

    swap(elements[mid], elements[high]); // Move median to pivot position
}

// Partition using the chosen pivot
int partition(int low, int high) {
    medianOfThree(low, high);
    int pivot = elements[high].first;
    int i = low - 1;

    for (int j = low; j < high; ++j) {
        if (elements[j].first < pivot) {
            ++i;
            swap(elements[i], elements[j]);
        }
    }
    swap(elements[i + 1], elements[high]);
    return i + 1;
}

// Optimized QuickSort
void quickSort(int low, int high) {
    while (low < high) {
        int pi = partition(low, high);

        // Tail call optimization: recurse into smaller partition
        if (pi - low < high - pi) {
            quickSort(low, pi - 1);
            low = pi + 1;
        } else {
            quickSort(pi + 1, high);
            high = pi - 1;
        }
    }
}

File: Assignment quick sort v2/newQuick.java


import java.io.*;
import java.util.*;
import java.nio.file.*;
import java.time.Duration;
import java.time.Instant;

public class newQuick {
    static List<Pair> elements = new ArrayList<>();
    //list to store each row from the CSV 

    public static void main(String[] args) {
        //input file to be read
        String inputFile = "dataset_5000000.csv";

        try (BufferedReader reader = Files.newBufferedReader(Paths.get(inputFile))) {
            String line;
            //read the file
            while ((line = reader.readLine()) != null) {
                //split into 2,number and text
                String[] tokens = line.split(",", 2);
                if (tokens.length == 2) {
                    try {
                        int number = Integer.parseInt(tokens[0]);
                        elements.add(new Pair(number, tokens[1]));
                    } catch (NumberFormatException e) {
                        System.err.println("Invalid line skipped: " + line);
                    }
                }
            }
        } catch (IOException e) {
            System.err.println("Cannot open file: " + inputFile);
            return;
        }
        //start time
        Instant start = Instant.now();
        quickSort(0, elements.size() - 1);
        //end time
        Instant end = Instant.now();

        //write output to csv file
        /*outputFile = "quick_sort_" + elements.size() + ".csv";

        try (BufferedWriter writer = Files.newBufferedWriter(Paths.get(outputFile))) {
            for (Pair p : elements) {
                writer.write(p.num + "," + p.text);
                writer.newLine();
            }
        } catch (IOException e) {
            System.err.println("Error writing to file: " + outputFile);
        }*/
        //calculate time taken
        Duration elapsed = Duration.between(start, end);
        System.out.println("Sorting time: " + elapsed.toMillis() / 1000.0 + " seconds");
    }

    // QuickSort with median-of-three
    public static void quickSort(int low, int high) {
        while (low < high) {
            int pi = partition(low, high);

            // Tail call optimization
            if (pi - low < high - pi) {
                quickSort(low, pi - 1);
                low = pi + 1;
            } else {
                quickSort(pi + 1, high);
                high = pi - 1;
            }
        }
    }

    public static int partition(int low, int high) {
        medianOfThree(low, high);
        int pivot = elements.get(high).num;
        int i = low - 1;

        for (int j = low; j < high; ++j) {
            if (elements.get(j).num < pivot) {
                i++;
                Collections.swap(elements, i, j);
            }
        }

        Collections.swap(elements, i + 1, high);
        return i + 1;
    }

    public static void medianOfThree(int low, int high) {
        int mid = low + (high - low) / 2;

        if (elements.get(high).num < elements.get(low).num)
            Collections.swap(elements, low, high);
        if (elements.get(mid).num < elements.get(low).num)
            Collections.swap(elements, mid, low);
        if (elements.get(high).num < elements.get(mid).num)
            Collections.swap(elements, mid, high);

        Collections.swap(elements, mid, high); // Move median to pivot position
    }

    // Pair class
    static class Pair {
        int num;
        String text;

        Pair(int num, String text) {
            this.num = num;
            this.text = text;
        }
    }
}

